unblob/handlers/filesystem/netgeartrx.py

```diff
@@ -0,0 +1,70 @@
+import io
+from pathlib import Path
+from typing import Optional
+
+from structlog import get_logger
+
+from unblob.extractor import carve_chunk_to_file
+from unblob.file_utils import Endian, File, StructParser
+from unblob.models import Chunk, Extractor, HexString, StructHandler, ValidChunk
+
+logger = get_logger()
+
+TRX_HEADER = r"""
+    typedef struct trx_header {
+        uint32 magic;              /* "HDR0" */
+        uint32 len;                /* Length of file including header */
+        uint32 crc32;              /* 32-bit CRC from flag_version to end of file */
+        uint32 flag_version;       /* 0:15 flags, 16:31 version */
+        uint32 offsets[3]; /* Offsets of partitions from start of header */
+    } trx_header_t;
+"""
+
+
+class TRXExtractor(Extractor):
+    def __init__(self):
+        self._struct_parser = StructParser(TRX_HEADER)
+
+    def extract(self, inpath: Path, outdir: Path):
+        with File.from_path(inpath) as file:
+            header = self._struct_parser.parse("trx_header_t", file, Endian.LITTLE)
+            file.seek(0, io.SEEK_END)
+            eof = file.tell()
+
+            self._dump_file(
+                file, outdir, Path("part0"), header.offsets[0], header.offsets[1], eof
+            )
+            self._dump_file(
+                file, outdir, Path("part1"), header.offsets[1], header.offsets[2], eof
+            )
+            self._dump_file(file, outdir, Path("part2"), header.offsets[2], 0, eof)
+
+    def _dump_file(
+        self, file: File, outdir: Path, path: Path, start: int, end: int, eof: int
+    ):
+        if not start:
+            return
+
+        if not end:
+            end = eof
+
+        chunk = Chunk(start_offset=start, end_offset=end)
+        carve_chunk_to_file(outdir.joinpath(path), file, chunk)
+
+
+class NetgearTRXHandler(StructHandler):
+    NAME = "trx"
+
+    PATTERNS = [HexString("48 44 52 30")]
+
+    C_DEFINITIONS = TRX_HEADER
+    HEADER_STRUCT = "trx_header_t"
+    EXTRACTOR = TRXExtractor()
+
+    def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]:
+        header = self.parse_header(file, endian=Endian.LITTLE)
```

Comment:
We usually recommend to add a header validation function to make sure we don't end up with false positives. A simple implementation woud look like this:

```python
from dissect.cstruct import Instance
# --snip---
def is_valid_header(header: Instance) -> bool:
    if header.len > len(header):
        return False
     return True
```

I would also look into implementing a CRC32 check since the format implements it. You can look at https://github.com/onekey-sec/unblob/blob/main/unblob/handlers/filesystem/cramfs.py#L57 for an example.

Reply:
Properly identifying if the CRC32 and header fields are valid is done.